### ISSUE TITLE
`jest.useFakeTimers()` stuck the test in MSW 2

### ISSUE THREAD

**Issue by joel-daros** [kettanaito/msw#1830]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 18 or higher

### Node.js version

v18.17.1

### Reproduction repository

https://github.com/joel-daros/msw2-text-encoder-issue

### Reproduction steps

Clone the repo and run:

```shell
npm install --legacy-peer-deps
npm run test
```

### Current behavior

After adding [this](https://github.com/joel-daros/msw2-text-encoder-issue/blob/ea2d0160b8f6daa958c218ec16ecb4b00bccb068/src/DogService.test.ts#L22-L24) line, the test stuck, and the only way to finishing is aborting the jest execution:

```ts
  // when this line is added, the test stuck and never finishes
  jest.useFakeTimers({ now: new Date(2023, 9, 15), doNotFake: ["setTimeout"] });
```

I noticed this issue has been addressed [here](https://github.com/mswjs/msw/issues/966) in MSW 1.x (and actually works well in 1.x version), but seems like it's back in MSW 2.x with the addition of the [jest.pollyfills.js](https://github.com/joel-daros/msw2-text-encoder-issue/blob/main/src/jest.polyfills.js) file.

### Expected behavior

`jest.useFakeTimers()` should be work in MSW 2.

**Comment by jeffwong14** [comment]
Looks similar issue happens to me in #1829 

My [Code Sandbox](https://codesandbox.io/p/sandbox/gifted-liskov-z5qwrn?file=%2Fpackage.json%3A6%2C19&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clohxk2dx00073b6ifydiey2m%2522%252C%2522sizes%2522%253A%255B70.27548209366391%252C29.72451790633609%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clohxk2dx00033b6i7vmf3df8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clohxk2dx00053b6i4u7of2mu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clohxk2dx00063b6iol903yi2%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clohxk2dx00033b6i7vmf3df8%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00033b6i7vmf3df8%2522%252C%2522activeTabId%2522%253A%2522clohzocjn016b3b6iiex808hd%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FgetData.test.ts%2522%252C%2522id%2522%253A%2522clohxrt8s00og3b6i4ke9g3aq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.config.js%2522%252C%2522id%2522%253A%2522clohziz9i00xu3b6ig0b7sk3z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.polyfills.js%2522%252C%2522id%2522%253A%2522clohzocjn016b3b6iiex808hd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.setup.js%2522%252C%2522id%2522%253A%2522clohzrho801v23b6iqz8h6trg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522clohzzq2403733b6i1a3fm5lw%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522clohxk2dx00063b6iol903yi2%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00063b6iol903yi2%2522%252C%2522activeTabId%2522%253A%2522clohxm93800cy3b6imhxhv61r%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clohxm93800cy3b6imhxhv61r%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522SYSTEM_METRICS%2522%252C%2522id%2522%253A%2522clohzpo1h01jj3b6it3oq3fiz%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clohxk2dx00053b6i4u7of2mu%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00053b6i4u7of2mu%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clohxkvk6006g3b6i3pfeync4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clohxkvna00dcefg66i1rgyc0%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clohxkvk6006g3b6i3pfeync4%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A26.27596439169139%257D) stucks at await response.json() if using fakeTimers

And the issue gone if I remove those fakeTimers code

**Comment by mattcosta7** [comment]
> Looks similar issue happens to me in [#1829](https://github.com/mswjs/msw/discussions/1829)
> 
> My [Code Sandbox](https://codesandbox.io/p/sandbox/gifted-liskov-z5qwrn?file=%2Fpackage.json%3A6%2C19&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clohxk2dx00073b6ifydiey2m%2522%252C%2522sizes%2522%253A%255B70.27548209366391%252C29.72451790633609%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clohxk2dx00033b6i7vmf3df8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clohxk2dx00053b6i4u7of2mu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clohxk2dx00063b6iol903yi2%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clohxk2dx00033b6i7vmf3df8%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00033b6i7vmf3df8%2522%252C%2522activeTabId%2522%253A%2522clohzocjn016b3b6iiex808hd%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FgetData.test.ts%2522%252C%2522id%2522%253A%2522clohxrt8s00og3b6i4ke9g3aq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.config.js%2522%252C%2522id%2522%253A%2522clohziz9i00xu3b6ig0b7sk3z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.polyfills.js%2522%252C%2522id%2522%253A%2522clohzocjn016b3b6iiex808hd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fjest.setup.js%2522%252C%2522id%2522%253A%2522clohzrho801v23b6iqz8h6trg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522clohzzq2403733b6i1a3fm5lw%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522clohxk2dx00063b6iol903yi2%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00063b6iol903yi2%2522%252C%2522activeTabId%2522%253A%2522clohxm93800cy3b6imhxhv61r%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clohxm93800cy3b6imhxhv61r%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522SYSTEM_METRICS%2522%252C%2522id%2522%253A%2522clohzpo1h01jj3b6it3oq3fiz%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clohxk2dx00053b6i4u7of2mu%2522%253A%257B%2522id%2522%253A%2522clohxk2dx00053b6i4u7of2mu%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clohxkvk6006g3b6i3pfeync4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clohxkvna00dcefg66i1rgyc0%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clohxkvk6006g3b6i3pfeync4%2522%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A26.27596439169139%257D) stucks at await response.json() if using fakeTimers
> 
> And the issue gone if I remove those fakeTimers code

A little attempt at some initial debugging:

```
    let x = resp.json();
    jest.runOnlyPendingTimers();
    x = await x;
```

I modified like this, to see if we could run it once jest progressed timers, and we can but this failed with a json parsing error.

> SyntaxError: Unexpected non-whitespace character after JSON at position 20


```
    let x = resp.text();
    jest.runOnlyPendingTimers();
    x = await x;
```

this shows a response that's wrong, which is interesting

```
{"result":"success"}{"result":"success"}
```

which doubles the response in the body.


(just leaving these breadcrumbs for debugging further)

**Comment by mattcosta7** [comment]
With `legacy fake timers` it appears to work properly:

```
  jest.useFakeTimers({
    now: hardCodeNow,
    legacyFakeTimers: true,
  });
//
    let x = await resp.json();
```

which I discovered as a setting from this: https://github.com/nock/nock/issues/2200

this config option in jest is 

```
  /**
   * Use the old fake timers implementation instead of one backed by `@sinonjs/fake-timers`.
   * The default is `false`.
   */
  legacyFakeTimers?: boolean;
```

which makes it seem like there's some issue related to `@sinonjs/fake-timers` somewhere, since it works with the other implementation, with no other changes.   It seems like they [recently started mocking](https://github.com/sinonjs/fake-timers/commit/a111a715c76542b15f4701ad8b1752789f597868#diff-41a23d75048c36f3649e9fea5cd63cab3d94802c03bc6df0fdc009847d786bf1) the `timers` module, and I wonder if this is causing issues somehow?

**Comment by mattcosta7** [comment]
Found another workaround you can see here

```ts
  jest.useFakeTimers({
    now: hardCodeNow,
    doNotFake: ["queueMicrotask"],
  });
```

It looks like it's related to this issue - https://github.com/nodejs/undici/issues/1251 in undici, which is really less of an undici issue and more of a "jest does some things in not great ways" issue.

As long as you don't need to fake calls to `queueMicrotask` this should be a viable workaround that maintains the 'modern' fake timer implementation


So in summary - I don't believe `msw` can directly solve this issue, but we can suggest that when using fake timers in jest the configuration be extended to either

1. use legacy fake times

```ts
  jest.useFakeTimers({
    now: hardCodeNow,
    legacyFakeTimers: true,
  });
```

2. avoid faking queueMicrotask

```ts
  jest.useFakeTimers({
    now: hardCodeNow,
    doNotFake: ["queueMicrotask"],
  });
```


Both of these configurations seem to work properly

**Comment by mattcosta7** [comment]
Since i'm not sure that msw can effectively fix this directly and there are at least 2 workarounds, I'm going to close this one - but please let me know if neither of those 2 help to resolve the issue which stems from how jest and undici (the node fetch library) interact.

**Comment by mattcosta7** [comment]
@kettanaito  - do you think it might be wise add some documentation here: https://mswjs.io/docs/runbook for this?  

If so (or if you think there's a better place) I'm happy to write up a draft

**Comment by kettanaito** [comment]
@mattcosta7, thanks for looking into this! Yes, I believe the Runbook is the best place for this. Please, would you have a minute to open a pull request with the `doNotFake` approach? I think we shouldn't influence whether the developer uses legacy/modern timers. Let them decide.

**Comment by mattcosta7** [comment]
> [@mattcosta7](https://github.com/mattcosta7), thanks for looking into this! Yes, I believe the Runbook is the best place for this. Please, would you have a minute to open a pull request with the `doNotFake` approach? I think we shouldn't influence whether the developer uses legacy/modern timers. Let them decide.

Opened https://github.com/mswjs/mswjs.io/pull/290

**Comment by joel-daros** [comment]
@mattcosta7 Thanks for the solution, but it seems that none of them are working in my reproduction repo in the first thread message. I didn't find any major differences in my testing approach compared to @jeffwong14's Codesandbox.

**Comment by mattcosta7** [comment]
> [@mattcosta7](https://github.com/mattcosta7) Thanks for the solution, but it seems that none of them are working in my reproduction repo in the first thread message. I didn't find any major differences in my testing approach compared to [@jeffwong14](https://github.com/jeffwong14)'s Codesandbox.

@joel-daros I haven't used create-react-app in a long time. From the [docs](https://create-react-app.dev/docs/running-tests/#srcsetuptestsjs), it looks like `src/setupTests` is configured to set `setupFilesAfterEnv` and not `setupFiles`.  

As we discuss in the docs for msw

> **warning**
> Pay attention it’s the setupFiles option, and not setupFilesAfterEnv. The missing Node.js globals must be injected before the environment (e.g. JSDOM).

I'm not sure it can be configured in create-react-app without ejecting (but you can always just use jest directly instead of via react-scripts and pass a custom configuration for jest that mirrors cra but with additional configuration 

https://github.com/facebook/create-react-app/blob/0a827f69ab0d2ee3871ba9b71350031d8a81b7ae/packages/react-scripts/scripts/utils/createJestConfig.js#L30C5-L35C1

so you may need to avoid using `react-scripts test` and manually configure jest to make this work - unless there's another way to override the actual `setupFiles` jest is using (maybe forking cra or using patch-package as 2 potential options)

**Comment by kettanaito** [comment]
I suppose `node --require ./setup.js node_modules/.bin/react-scripts test` should work too. Depends on what `react-scripts` actually does process-wise.

**Comment by mattcosta7** [comment]
> I suppose `node --require ./setup.js node_modules/.bin/react-scripts test` should work too. Depends on what `react-scripts` actually does process-wise.

Maybe!  I think it might conflict on `react-app-polyfill/jsdom` which will polyfill an alternative version of fetch, and I imagine the competing fetch/request/response impls will be problematic

**Comment by michalstrzelecki** [comment]
@mattcosta7 2nd fixed the issue, I have not checked 1st.
